### ISSUE TITLE
Exec into Running Pods instead of Terminating ones

### ISSUE THREAD

**Issue by cawoodm** [kubernetes/kubectl#1370]
**What happened**:
`kubectl exec` consistently runs commands in terminating pods instead of running ones.

**What you expected to happen**:
I would expect exec to prefer running pods and not exec into terminating pods at all.
At the very least it would be good to be able to control this behaviour as many people need to `exec` into fresh pods to run tests.
The current behaviour as described [here](https://stackoverflow.com/questions/75436733/how-to-kubectl-exec-to-freshly-created-pod) and [here](https://stackoverflow.com/questions/73389359/kubectl-rollout-status-does-not-wait-for-deployment-to-complete-and-pod-to-run) makes this impossible.

**How to reproduce it (as minimally and precisely as possible)**:
Consider a deployment with versioned tag `1.0.0`. In order to deploy a new version v2 we run:
```bash
kubectl set image deploy/myapp myapp=myapp:2.0.0 && kubectl rollout status deploy/myapp && check1
```
Where `check1` is an alias which runs `curl localhost:8080/ping` which returns a BUILD_ID:
```bash
alias check1="date; kubectl get deploy/myapp; kubectl get pod -l app=myapp; kubectl exec deploy/myapp -- curl localhost:8080/ping -sS | grep BUILD_ID"
// Would expect 2.0.0 here but get 1.0.0
```
After some time (20-30s) we eventually get `check1` to return `2.0.0`. This is strange because our app doesn't really need much time to shut down.

Now we deploy version 1 again:
```bash
k set image deploy/myapp myapp=myapp:2.0.0 && kubectl rollout status deploy/myapp && check1
// Would expect 1.0.0 here but get 2.0.0
```
As you will see the check1 shows pods running and ready but the curl statement shows the old version running.

**Anything else we need to know?**:

**Environment**:

### Client Version:
```
{Major:"1", Minor:"23", GitVersion:"v1.23.6+k3s1", GitCommit:"418c3fa858b69b12b9cefbcff0526f666a6236b9", GitTreeState:"clean", BuildDate:"2022-04-28T22:16:18Z", GoVersion:"go1.17.5", Compiler:"gc", Platform:"linux/amd64"}
Server Version: version.Info{Major:"1", Minor:"23", GitVersion:"v1.23.6+k3s1", GitCommit:"418c3fa858b69b12b9cefbcff0526f666a6236b9", GitTreeState:"clean", BuildDate:"2022-04-28T22:16:18Z", GoVersion:"go1.17.5", Compiler:"gc", Platform:"linux/amd64"}
```

### OS:
```
NAME="Ubuntu"
VERSION="20.04.5 LTS (Focal Fossa)"
ID=ubuntu
ID_LIKE=debian
PRETTY_NAME="Ubuntu 20.04.5 LTS"
VERSION_ID="20.04"
HOME_URL="https://www.ubuntu.com/"
SUPPORT_URL="https://help.ubuntu.com/"
BUG_REPORT_URL="https://bugs.launchpad.net/ubuntu/"
PRIVACY_POLICY_URL="https://www.ubuntu.com/legal/terms-and-policies/privacy-policy"
VERSION_CODENAME=focal
UBUNTU_CODENAME=focal
```

**Comment by cawoodm** [comment]
According to [this answer](https://stackoverflow.com/a/75443123/279255) a terminating pod is still running and thus a candidate for exec. This may be intended behaviour but it's a problem for such deployment scenarios described.

If we reduce `terminationGracePeriodSeconds` to 3 the process is faster but it's not clear why Kubernetes is waiting the full 30s before deleting the pod.

**Comment by nitishchauhan0022** [comment]
@sftim @cawoodm PTAL

**Comment by ardaguclu** [comment]
I think this issue is fixed by https://github.com/kubernetes/kubernetes/pull/119256 and I think we can close. @brianpursley am I correct?

**Comment by brianpursley** [comment]
Yeah, good catch @ardaguclu 

I think this was the same problem, so it should be fixed in kubectl 1.28.

/close
